File: DiffViewer.java
// DiffViewer.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import difflib.*;

public class DiffViewer {
    private static boolean color = true;

    public static void main(String[] args) throws Exception {
        String file1 = null, file2 = null, output = null;
        int context = 3;
        boolean ignoreSpace = false;
        boolean unified = true;
        boolean forceColor = false;

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--context": context = Integer.parseInt(args[++i]); break;
                case "--ignore-space": ignoreSpace = true; break;
                case "--unified": unified = Boolean.parseBoolean(args[++i]); break;
                case "--color": forceColor = true; break;
                case "--output": output = args[++i]; break;
                case "--help":
                    System.out.println("Usage: java DiffViewer <file1> <file2> [options]");
                    return;
                default:
                    if (file1 == null) file1 = args[i];
                    else if (file2 == null) file2 = args[i];
                    else System.err.println("Unknown argument: " + args[i]);
            }
        }

        if (file1 == null || file2 == null) {
            System.err.println("Two files are required.");
            System.exit(1);
        }

        color = forceColor || System.console() != null;

        List<String> lines1 = Files.readAllLines(Paths.get(file1));
        List<String> lines2 = Files.readAllLines(Paths.get(file2));
        if (ignoreSpace) {
            lines1 = lines1.stream().map(s -> s.replaceAll("\\s+$", "")).toList();
            lines2 = lines2.stream().map(s -> s.replaceAll("\\s+$", "")).toList();
        }

        Patch patch = DiffUtils.diff(lines1, lines2);
        List<String> result = new ArrayList<>();
        result.add(colorize("--- " + file1, "blue"));
        result.add(colorize("+++ " + file2, "blue"));

        for (Delta delta : patch.getDeltas()) {
            // Упрощённый вывод: показываем удалённые и добавленные строки
            List<String> linesDel = delta.getOriginal().getLines();
            List<String> linesAdd = delta.getRevised().getLines();
            for (String l : linesDel) {
                result.add(colorize("-" + l, "red"));
            }
            for (String l : linesAdd) {
                result.add(colorize("+" + l, "green"));
            }
        }

        String out = String.join("\n", result);
        if (output != null) {
            Files.write(Paths.get(output), out.getBytes());
            System.out.println("Diff saved to " + output);
        } else {
            System.out.println(out);
        }
    }

    private static String colorize(String text, String colorName) {
        if (!color) return text;
        Map<String, String> codes = new HashMap<>();
        codes.put("red", "\033[91m");
        codes.put("green", "\033[92m");
        codes.put("blue", "\033[94m");
        codes.put("reset", "\033[0m");
        return codes.get(colorName) + text + codes.get("reset");
    }
}

File: diff_viewer.py
# diff_viewer.py
import sys
import argparse
import difflib
import os
from datetime import datetime

# ANSI color codes
COLORS = {
    'red': '\033[91m',
    'green': '\033[92m',
    'blue': '\033[94m',
    'reset': '\033[0m'
}

def colorize(text, color, enabled):
    if not enabled:
        return text
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

def read_file(path):
    try:
        with open(path, 'r', encoding='utf-8') as f:
            return f.read().splitlines()
    except Exception as e:
        print(f"Error reading {path}: {e}", file=sys.stderr)
        sys.exit(1)

def main():
    parser = argparse.ArgumentParser(description="Diff Viewer")
    parser.add_argument('file1', help='First file')
    parser.add_argument('file2', help='Second file')
    parser.add_argument('--context', type=int, default=3, help='Context lines')
    parser.add_argument('--ignore-space', action='store_true', help='Ignore trailing spaces')
    parser.add_argument('--unified', action='store_true', default=True, help='Unified format')
    parser.add_argument('--color', action='store_true', help='Force color output')
    parser.add_argument('--output', help='Output file')
    args = parser.parse_args()

    color = args.color or sys.stdout.isatty()
    lines1 = read_file(args.file1)
    lines2 = read_file(args.file2)

    # Подготовка
    if args.ignore_space:
        lines1 = [line.rstrip() for line in lines1]
        lines2 = [line.rstrip() for line in lines2]

    # Генерируем diff
    diff = difflib.unified_diff(
        lines1, lines2,
        fromfile=args.file1, tofile=args.file2,
        lineterm='',
        n=args.context
    )
    diff_lines = list(diff)

    if not diff_lines:
        print("No differences found.")
        return

    # Форматируем вывод с цветами
    output_lines = []
    for line in diff_lines:
        if line.startswith('---') or line.startswith('+++') or line.startswith('@@'):
            output_lines.append(colorize(line, 'blue', color))
        elif line.startswith('-'):
            output_lines.append(colorize(line, 'red', color))
        elif line.startswith('+'):
            output_lines.append(colorize(line, 'green', color))
        else:
            output_lines.append(line)

    if args.output:
        with open(args.output, 'w', encoding='utf-8') as f:
            f.write('\n'.join(output_lines))
        print(f"Diff saved to {args.output}")
    else:
        print('\n'.join(output_lines))

if __name__ == '__main__':
    main()
